### DISCUSSION TITLE
Pixel-perfect screenshot tests on different environments

### DISCUSSION THREAD

**Discussion by SimonMarquis** [android/nowinandroid#1646] [General]
While having pixel-perfect screenshots seems a good idea, it makes tests fail almost certainly on environments that are not 100% exact to the CI env.
Could we lower our expectations and restore the default `changeThreshold` (`0.01F`) in `RoborazziOptions`?

https://github.com/android/nowinandroid/blob/c4ebde1fc22d2232375046e9fe43350f29b70e6e/core/screenshot-testing/src/main/kotlin/com/google/samples/apps/nowinandroid/core/testing/util/ScreenshotHelper.kt#L42-L43

For example, running on Windows with `jbr-17.0.7` leads to this output:

```
Roborazzi: GradientBackground_light_androidTheme_notDynamic.png The differ result :ComparisonResult(pixelDifferences=4, pixelCount=22500, width=150, height=150) changed:false
Roborazzi: GradientBackground_light_defaultTheme_dynamic.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=22500, width=150, height=150) changed:false
Roborazzi: Background_light_androidTheme_notDynamic.png The differ result :ComparisonResult(pixelDifferences=4, pixelCount=22500, width=150, height=150) changed:false
Roborazzi: FilterChipSelected_dark_androidTheme_notDynamic.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=12744, width=177, height=72) changed:false
Roborazzi: Navigation_fontScale2.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=73440, width=480, height=153) changed:false
Roborazzi: TopAppBar_fontScale2.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=46080, width=480, height=96) changed:false
Roborazzi: ForYouScreenPopulatedAndLoading_foldable.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=1175485, width=1009, height=1165) changed:false
Roborazzi: ForYouScreenPopulatedAndLoading_tablet.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=2119680, width=1920, height=1104) changed:false
Roborazzi: ForYouScreenPopulatedFeed_foldable.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=1175485, width=1009, height=1165) changed:false
Roborazzi: ForYouScreenPopulatedFeed_tablet.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=2119680, width=1920, height=1104) changed:false
Roborazzi: insets_snackbar_medium_medium.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=810000, width=900, height=900) changed:false
Roborazzi: insets_snackbar_expanded_expanded.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=1822500, width=1350, height=1350) changed:false
Roborazzi: snackbar_medium_medium.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=810000, width=900, height=900) changed:false
Roborazzi: snackbar_expanded_expanded.png The differ result :ComparisonResult(pixelDifferences=1, pixelCount=1822500, width=1350, height=1350) changed:false
```

As you can see, the `pixelDifferences` is generally very low, and we should probably allow some very slight variation, while still keeping the CI generated golden images as the single source of truth.

PS: There might be a good reason for it, but I did not find any mention of why pixel-perfect screenshots were configured in the first place: https://github.com/android/nowinandroid/pull/876/files#diff-4d54482e44efb1225f70ed9435e0c5da894e62592a8058f607ef7d3e22ba9e6dR35 If it is for CI purposes only, maybe we could switch behaviors based on env variables.

**Comment by JoseAlcerreca** [comment]
We create screenshots on CI to avoid this issue. Adding a threshold is dangerous because you start with a small one and there's always a reason–bigger screenshots–to increase it. You end up accepting diffs that should have been a fail in small components.
